programs/ui-wrapper/src/processors/place_order.rs

```diff
@@ -0,0 +1,260 @@
+    let vault: &AccountInfo = next_account_info(account_iter)?;
+    let mint: &AccountInfo = next_account_info(account_iter)?;
+    let system_program: Program =
+        Program::new(next_account_info(account_iter)?, &system_program::id())?;
+    let token_program: &AccountInfo = next_account_info(account_iter)?;
+    let manifest_program: Program =
+        Program::new(next_account_info(account_iter)?, &manifest::id())?;
+    let payer: Signer = Signer::new(next_account_info(account_iter)?)?;
+
+    if spl_token_2022::id() == *token_program.key {
+        unimplemented!("token2022 not yet supported")
+    }
+
+    check_signer(&wrapper_state, owner.key);
+    let market_info_index: DataIndex = get_market_info_index_for_market(&wrapper_state, market.key);
+
+    // Do an initial sync to get all existing orders and balances fresh. This is
+    // needed for modifying user orders for insufficient funds.
+    sync_fast(&wrapper_state, &market, market_info_index)?;
+
+    let order = WrapperPlaceOrderParams::try_from_slice(data)?;
+
+    let wrapper_data: Ref<&mut [u8]> = wrapper_state.info.try_borrow_data()?;
+    let (_fixed_data, wrapper_dynamic_data) =
+        wrapper_data.split_at(size_of::<ManifestWrapperStateFixed>());
+
+    let market_info: MarketInfo =
+        *get_helper::<RBNode<MarketInfo>>(wrapper_dynamic_data, market_info_index).get_value();
+    let remaining_base_atoms: BaseAtoms = market_info.base_balance;
+    let remaining_quote_atoms: QuoteAtoms = market_info.quote_balance;
+    let trader_index: DataIndex = market_info.trader_index;
+    drop(wrapper_data);
+
+    let base_atoms = BaseAtoms::new(order.base_atoms);
+    let price = QuoteAtomsPerBaseAtom::try_from_mantissa_and_exponent(
+        order.price_mantissa,
+        order.price_exponent,
+    )?;
+
+    let deposit_amount: u64 = if order.is_bid {
+        let required_quote_atoms = base_atoms.checked_mul(price, true)?;
+        required_quote_atoms
+            .saturating_sub(remaining_quote_atoms)
+            .as_u64()
+    } else {
+        base_atoms.saturating_sub(remaining_base_atoms).as_u64()
+    };
+
+    trace!("deposit amount:{deposit_amount} mint:{:?}", mint.key);
+    if deposit_amount > 0 {
+        invoke(
+            &deposit_instruction(
+                market.key,
+                owner.key,
+                mint.key,
+                deposit_amount,
+                trader_token_account.key,
+                *token_program.key,
+            ),
+            &[
+                manifest_program.info.clone(),
+                owner.info.clone(),
+                market.info.clone(),
+                trader_token_account.clone(),
+                vault.clone(),
+                token_program.clone(),
+                mint.clone(),
+            ],
+        )?;
+    }
+
+    // Call expand so claim seat has enough free space
+    // and owner doesn't get charged rent
+    // TODO: could check if needed before
```

Comment:
This is probably needed since it will be called a lot.
Should be a reasonably quick check to verify that free_list_head_index != NIL

Reply:
i actually need to check if the freelist has 2 items at least, bc. of the last block free policy

Reply:
freelist is unfortunately private as well, imo something to keep as TODO